Check for common mistake in storage instances/ endpoint

ivarne opened the pull request:
The [`instances/` api in storage ](https://docs.altinn.studio/api/storage/instances/#query-instances) requires `appId` to be on the format `org/appname`. It's easy for api users to get confused when the docs has fields for both `org` and `appId` (as I was), and this small change improves the request validation to ensure that `BadRequest` is returned if the request is bad.


## Related Issue(s)
- https://altinn.slack.com/archives/C02EVE480G2/p1654580020477629

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

SandGrainOne commented:
I wish we could merge this, but it's a major breaking change. I would assume most if not all existing API users are using both parameters today. Necessary or not.

ivarne commented:
@SandGrainOne I don't think I wrote braking code (except for inconsistent requests that previously returned an empty list).

The relevant part of this discussion is how to react when `org` and `appId` is defined or undefined, as can be summed up in a truth table

```
                 | app defined              | app defined    | app defined   | app undefined
                 | (correct org and format) | (wrong format) | (wrong org)   |
---------------- | ------------------------ | -------------- | ------------- | --------------
org defined      |             1.           |       2.       |       3.      |      4.
org undefined    |             5.           |       6.       |       7.      |      8.

```
The previous behaviour was as the following
1. Correct results returned
2. Empty list of instances
3. Empty list or Forbid (auth can only be for a single org)
4. All instances for the org?
5. Org overwritten with the org part of `org/appName` in `appId`
6. Forbid (user has not access to org `krt-1234`)
7. Org can't be wrong when not specified
8. BadRequest

After my changes 
1. ..
2. BadRequest("appId must be on the format org/appname.")
3. BadRequest("appId must be on the format org/appname.")
4. ...
5. ...
6. BadRequest("appId must be on the format org/appname.")
7. ...
8. ...

**Note:** I tested with specifying both `org` and `appId` = `org/appname`, and there is no change there.

SandGrainOne commented:
You're right. I had an idea in my head and made assumptions about the code.

SandGrainOne commented on src/Altinn.Platform/Altinn.Platform.Storage/Storage/Controllers/InstancesController.cs:
```diff
@@ -146,10 +146,18 @@ public class InstancesController : ControllerBase
+                if (!string.IsNullOrEmpty(appId))
+                {
+                    var appParts = appId.Split('/', 2);
+                    if(appParts.Length != 2 || (string.IsNullOrEmpty(org) && appParts[0] != org))
```
Should this have been _not null or empty_?

ivarne replied:
I should have a pre commit hook that issues a warning when i reach for `isset` and forget to write the `!` in front of `string.isNullOrEmpty`.
```suggestion
                    if(appParts.Length != 2 || (!string.IsNullOrEmpty(org) && appParts[0] != org))
```

SandGrainOne commented:
Sorry for taking so long here. A lot of activity because of VAT reporting deadline today.

I've identified an issue with a test. The test `GetMany_OrgRequestsInstancesNoOrgDefined_ReturnsBadRequest` must be updated with a new expected value. Could you please look into that. I'll see if I can write a few other tests for the new case when I have time.

ivarne commented:
Yea, the test simply fails because I changed the capitalization of the error message to match the capitalization of the documented parameters. I'm also violating some coding standard conventions, so I'm surprised that the CI did not catch any of that.

I'm looking into writing some tests, and came to the conclusion that this controller method is complicated enough to deserve a set of unit tests to ensure that all the auth related logic is correct, and that rewriting does not change behaviour. I'm currently stuck because I need to pass all the arguments as parameters in addition to as a query string. It's defiantly on my list, but that list seems to be growing faster than I can finish anything.

ivarne commented:
@SandGrainOne I have moved the logic so that i applies to user authentication. I also added unit tests for the `InstancesController.Get`, to help ensure that I don't miss `!` in some critical places. I have also copied the logic to `LocalTest`, and updated a few other files to mirror changes done in `Altinn.Platform.Storage`.

SandGrainOne commented on src/Altinn.Platform/Altinn.Platform.Storage/UnitTest/TestingControllers/InstancesControllerGETTests.cs:
```diff
@@ -0,0 +1,340 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+using NSubstitute;
```
All our test projects already have references to Moq and I would prefer if that was what we continued to use.

ivarne replied:
Done!

ivarne commented on src/Altinn.Platform/Altinn.Platform.Storage/Storage/Controllers/InstancesController.cs:
```diff
@@ -173,7 +195,24 @@ public class InstancesController : ControllerBase
+                { "status.isSoftDeleted", isSoftDeleted ? "true" : null },
+                { "status.isHardDeleted", isHardDeleted ? "true" : null },
+                { "status.isArchived", isArchived ? "true" : null },
+            }.Where(el => el.Value != StringValues.Empty).ToDictionary(x => x.Key, x => x.Value);
```
Ooops! I just noticed that this approach will break a query on the form `dueBefore=gt:2019-02&dueBefore=lt:2019-03-01`. I've never seen that pattern before.

Maybe it can be fixed by changing the type signature for the dateTime query elements to `StringValues`, and maybe the API docs will match the API better too. The very minimum would be integration tests that ensures that date time range queries work.

SandGrainOne replied:
`instanceOwner.PartyId` is another example where we currently want to support multiple values through repeated use of the same parameter. 

For now I would like to see this particular change reverted.

lorang92 commented:
@ivarne I see this has gone a bit stale. Are you currently working on this? Or can we close and reopen when the work resumes? 👼

nkylstad commented:
@ivarne Any updates here? Can we close this PR?

nkylstad commented:
There have been no updates here in 4 months. Closing this PR, re-open if it's closed in error.
